core/jobs/batch_jobs/manual_voice_artist_name_job.py

```diff
@@ -0,0 +1,506 @@
+
+        Returns:
+            dict(str, dict(str, VoiceoverDict)). A dictionary maps
+            content IDs as keys and nested dicts as values. Each nested dict
+            contains language codes as keys and voiceover dicts as values. The
+            dictionary representing the difference between the old and new
+            values of the recorded voiceover.
+        """
+
+        new_voiceover_mapping = {}
+        old_voiceover_mapping = {}
+
+        for state in new_snapshot.content['states'].values():
+            voiceover_mapping = (
+                state['recorded_voiceovers']['voiceovers_mapping'])
+            new_voiceover_mapping.update(voiceover_mapping)
+
+        for state in old_snapshot.content['states'].values():
+            voiceover_mapping = (
+                state['recorded_voiceovers']['voiceovers_mapping'])
+            old_voiceover_mapping.update(voiceover_mapping)
+
+        voiceover_mapping_diff = (
+            self.get_voiceover_from_recorded_voiceover_diff(
+                new_voiceover_mapping=new_voiceover_mapping,
+                old_voiceover_mapping=old_voiceover_mapping
+            )
+        )
+        return voiceover_mapping_diff
+
+    def get_content_id_to_voice_artist_mapping(
+        self,
+        exploration: exp_models.ExplorationModel
+    ) -> voiceover_models.ContentIdToVoiceoverMappingType:
+        """This function generates a dictionary based on the contents of the
+        provided exploration and associates it with the relevant voice artist
+        ID responsible for the voiceovers.
+
+        Args:
+            exploration: ExplorationModel. The exploration model from which the
+                dict will be created.
+
+        Returns:
+            ContentIdToVoiceoverMappingType. A dictionary with content IDs as
+            keys and nested dicts as values. Each nested dict contains language
+            codes as keys and a 2-tuple as values. The 2-tuple contains voice
+            artist ID as the first element and VoiceoverDict as the second
+            element.
+        """
+
+        content_id_to_voiceovers_mapping: (
+            voiceover_models.ContentIdToVoiceoverMappingType) = {}
+        for state in exploration.states.values():
+            voiceover_mapping = (
+                state['recorded_voiceovers']['voiceovers_mapping']
+            )
+            for content_id, lang_code_to_voiceovers in (
+                    voiceover_mapping.items()):
+
+                content_id_to_voiceovers_mapping[content_id] = {}
+
+                for lang_code, voiceover_dict in (
+                        lang_code_to_voiceovers.items()):
+
+                    # Empty strings will be assigned as the voice artist name is
+                    # currently unknown, but it will be updated during iteration
+                    # on snapshot models.
+                    content_id_to_voiceovers_mapping[content_id][lang_code] = (
+                        '', voiceover_dict
```

Comment:
Use None rather than empty string to denote a null/non-existent value.

In general do not use the empty string as a sentinel. Null is meant for that purpose.

I am also a bit concerned about how the typing is handled. It seems to me like you are using an empty string because you want to keep the type consistent with the eventual type. But what you are doing here (assigning empty string and then populating some/all of it later) is a bit of a hack, and makes it hard to tell what the actual state is at any given time, and doesn't provide assurance that the end result only consists of non-empty strings. 

Instead, why don't you have this method return just `content_id_to_voiceovers_mapping[content_id][lang_code] = voiceover_dict`, and then pass the result into another method which also takes all the revisions of the exploration and does a for loop through them, incrementally building a new final object that always uses (non-empty str, voiceover_dict) pairs? Then you would never need to deal with the `''` case.